handle chain reorgs by refetching events

schmir opened the pull request:
we consider events/blocks from the chain final if they had
Synchronizer.required_confirmations confirmations. When the chain moves forward,
we fetch events for all blocks that were not considered final till the end of
the chain.

If blocks come in one by one, this means we're fetching events for each block
required_confirmations times. This may become something we have to optimize.

In that case take a look at the other solution I proposed in
https://github.com/trustlines-network/py-eth-index/pull/28 (exclude the last
commit), which would only fetch events for each block twice at most. The
disadvantage being that we may have some inconsistencies in the last blocks and
the chain reorg detection would still have to implemented.

Anyway, this solution here is a bit rude, but it is much simpler.

see issue https://github.com/trustlines-network/py-eth-index/issues/4 and
https://github.com/trustlines-network/py-eth-index/pull/28 for my previous
attempt.

schmir committed:
handle chain reorgs by refetching events

we consider events/blocks from the chain final if they had
Synchronizer.required_confirmations confirmations. When the chain moves forward,
we fetch events for all blocks that were not considered final till the end of
the chain.

If blocks come in one by one, this means we're fetching events for each block
required_confirmations times. This may become something we have to optimize.

In that case take a look at the other solution I proposed in
https://github.com/trustlines-network/py-eth-index/pull/28 (exclude the last
commit), which would only fetch events for each block twice at most. The
disadvantage being that we may have some inconsistencies in the last blocks and
the chain reorg detection would still have to implemented.

Anyway, this solution here is a bit rude, but it is much simpler.

see issue https://github.com/trustlines-network/py-eth-index/issues/4 and
https://github.com/trustlines-network/py-eth-index/pull/28 for my previous
attempt.

schmir committed:
get rid of the unused 'last_block_hash' field

We introduced a change to the database schema anyway, so let's get rid of that.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/trustlines-network/py-eth-index?pullRequest=37) <br/>All committers have signed the CLA.

schmir committed:
update changelog, next version will be 0.2.0

database schema has changed, so rebuild database

schmir committed:
add runsync --required-confirmations option

schmir committed:
replace add_some_tranfer_events with event_emitter fixture

add_some_tranfer_events was a 'side-effect only' fixture, that addes some events
to the blockchain. This was a bit confusing since in the test function this
fixture wasn't used.

Also adds ethereum_tester as a field to TestEnv. This is meant as a preparation
to write tests for the chain reorg handling.

fixes https://github.com/trustlines-network/py-eth-index/issues/36

schmir committed:
test that chain reorgs are handled
